File: Projeto e Analise de Algoritmos/Trabalho 3/c/heap.c
/*
* HEAP MIN - BASIC FUNCTIONS
*/

#include <stdio.h>
#include <stdlib.h>


int issetNode(int node, int *size){
	return (node < *size);
}

int getFather(int *heap, int node){
	return heap[(int) node/2];
}

int getLeftSon(int *heap, int node, int *size){
	node = node*2;
	if(issetNode(node, &(*size)))
		return heap[node];
	return -1;
}

int getRightSon(int *heap, int node, int *size){
	node = node*2 + 1;
	if(issetNode(node, &(*size)))
		return heap[node];
	return -1;
}

void rollUp(int *heap, int node, int *size){
	int temp, pos;
	while(getFather(heap, node) > heap[node]){
		temp = getFather(heap, node);
		pos = node/2;
		heap[pos] = heap[node];
		heap[node] = temp;
		node = pos;
	}
}

void rollDown(int *heap, int node, int *size){
	int temp, pos;
	while(heap[node] > getLeftSon(heap, node, &(*size)) || heap[node] > getRightSon(heap, node, &(*size))){
		if(getRightSon(heap, node, &(*size)) < getLeftSon(heap, node, &(*size)))
			pos = node*2 + 1;
		else
			pos = node*2;

		if(!issetNode(node, &(*size)))
			break;

		temp = heap[pos];
		heap[pos] = heap[node];
		heap[node] = temp;
		node = pos;
	}
}


void swap(int *heap, int node1, int node2){
	int temp = heap[node1];
	heap[node1] = heap[node2];
	heap[node2] = temp;
}


void heapify(int *heap, int node, int *size){
	int bigger = node;

	if(getLeftSon(heap, node, &(*size)) < heap[node] && getLeftSon(heap, node, &(*size)) != -1)
		bigger = node*2;

	if(getRightSon(heap, node, &(*size)) < heap[bigger] && getRightSon(heap, node, &(*size)) != -1)
		bigger = node*2 + 1;

	if(bigger != node){
		swap(heap, node, bigger);
		heapify(heap, bigger, &(*size));
	}
}


void buildHeap(int *heap, int *size){
	int i;
	for(i = (*size)/2; i>0; --i)
		heapify(heap, i, &(*size));
}

void insertNodeWithoutRoll(int *heap, int node, int *size, int *capacity){
	int pos = (*size);
	if(*size + 1 > *capacity){
		printf("Numero maximo de nodos no heap!\n");
		exit(0);
	}
	
	heap[pos] = node;
	(*size) += 1;
}


int extractMin(int *heap, int *size){
	int temp = heap[1];
	int pos = (*size)-1;
	heap[1] = heap[pos];
	*size -= 1;
	rollDown(heap, 1, &(*size));
	return temp;
}


void insertNode(int *heap, int node, int *size, int *capacity){
	int pos = (*size);
	if(*size + 1 > *capacity){
		printf("Numero maximo de nodos no heap!\n");
		exit(0);
	}
	
	heap[pos] = node;
	(*size) += 1;
	rollUp(heap, pos, &(*size));
}


void printHeap(int *heap, int *size){
	int i=1, nl=2;
	while(i < (*size)){
		printf("%d ", heap[i]);
		i++;
		if(i == nl){
			nl *= 2;
			printf("\n");
		}
	}
}



void iniatlize(int *heap, int *size){
	*size = 1;
	heap[0] = -1;
}


int main(int argc, char **argv){

	if(argc < 2){
        printf("\nErro! Informe a capacidade maxima do heap.\n");
        return 1;
    }

	int capacity = atoi(argv[1]);
	int *heap = (int*) malloc (capacity*sizeof(int));
	int size;

	iniatlize(heap, &size);

	insertNode(heap, 3, &size, &capacity);
	insertNode(heap, 7, &size, &capacity);
	insertNode(heap, 5, &size, &capacity);
	insertNode(heap, 8, &size, &capacity);
	insertNode(heap, 13, &size, &capacity);
	insertNode(heap, 6, &size, &capacity);
	insertNode(heap, 12, &size, &capacity);
	insertNode(heap, 15, &size, &capacity);
	insertNode(heap, 8, &size, &capacity);

	insertNode(heap, 4, &size, &capacity);

	printHeap(heap, &size);
	printf("\n");

	int min = extractMin(heap, &size);
	printf("\nMIN: %d\n", min);

	printHeap(heap, &size);

	printf("\n");
	return 0;
}

File: Projeto e Analise de Algoritmos/Trabalho 3/python/heap.py
class HeapMin:

	__heap = []
	__capacity = 0
	__inf = float("inf")

	def __init__(self, node, capacity, inf):
		self.__heap.append(node)
		self.__capacity = capacity
		self.__inf = inf


	def __issetNode(self, node):
		return (node < len(self.__heap))


	def __getFather(self, node):
		return self.__heap[int(node/2)]


	def __getLeftSon(self, node):
		if(self.__issetNode(node*2)):
			return self.__heap[node*2]
		return self.__inf


	def __getRightSon(self, node):
		if(self.__issetNode(node*2+1)):
			return self.__heap[node*2 + 1]
		return self.__inf


	def __rollUp(self, node):
		while self.__getFather(node).freq > self.__heap[node].freq:
			temp = self.__getFather(node)
			self.__heap[int(node/2)] = self.__heap[node]
			self.__heap[node] = temp;
			node = int(node/2)


	def __rollDown(self, node):
		while self.__heap[node].freq > self.__getLeftSon(node).freq or self.__heap[node].freq > self.__getRightSon(node).freq:
			
			if(self.__getRightSon(node).freq < self.__getLeftSon(node).freq):
				pos = node*2 + 1
			else:
				pos = node*2

			if(self.__issetNode(pos) == False):
				break

			temp = self.__heap[pos]
			self.__heap[pos] = self.__heap[node]
			self.__heap[node] = temp
			node = pos


	def __swap(self, node1, node2):
		temp = self.__heap[node1]
		self.__heap[node1] = self.__heap[node2]
		self.__heap[node2] = temp


	def heapify(self, node):
		bigger = node

		if(self.__getLeftSon(node).freq < self.__heap[node].freq and self.__getLeftSon(node) != False):
			bigger = node*2

		if(self.__getRightSon(node).freq < self.__heap[bigger].freq and self.__getRightSon(node) != False):
			bigger = node*2 + 1

		if(bigger != node):
			self.__swap(node, bigger)
			self.heapify(bigger)


	def buildHeap(self):
		i = int(len(self.__heap)/2);
		while(i>0):
			self.heapify(i)
			i-=1


	def insertNodeWithoutRoll(self, node):
		if(len(self.__heap)+1 > self.__capacity):
			raise CapacityError("Reached Capacity", "Maximum numbers of nodes in heap was reached!")

		self.__heap.append(node)


	def extractMin(self):
		if(len(self.__heap) > 2):
			temp = self.__heap[1]
			self.__heap[1] = self.__heap.pop()
			self.__rollDown(1)
		else:
			temp = self.__heap.pop()

		return temp


	def insertNode(self, node):
		pos = len(self.__heap)
		if(pos+1 > self.__capacity):
			raise CapacityError("Reached Capacity", "Maximum numbers of nodes in heap was reached!")

		self.__heap.append(node)
		self.__rollUp(pos)


	def printHeap(self):
		i, nl = 1, 2

		while(i < len(self.__heap)):
			print(self.__heap[i], end=" ")
			i += 1

			if(i == nl):
				nl *= 2
				print("\n", end="")


	def getSize(self):
		return len(self.__heap)


	def getNode(self, pos):
		return self.__heap[pos]



class CapacityError(Exception):
    def __init__(self, expr, msg):
        self.expr = expr
        self.msg = msg